### ISSUE TITLE
What coordinates should be handed to visHeatmap?

### ISSUE THREAD

**Issue by stephandooper** [mahmoodlab/CLAM#41]
I was wondering what coordinates have to be used as input to the visHeatmap function defined in the WholeSlideImage class. 
Are these the same coordinates that are present in the h5 files? If so, does it matter at which level of the pyramid you samples from (e.g. level 1 instead of level 0).

Finally, is it also possible to create fine attention maps from this function?

**Comment by fedshyvana** [comment]
Yes, I believe the coordinates are the same - it should be the coordinates of each patch relative the highest pyramid level (level 0). It doesn't matter which level you sampled patches from, you can visualize the heatmap at any level you want as long as you specify the correct patch_size. e.g. if you took 256 x 256 patches from level 0, your patch_size would be (256, 256), but if you took 256 x 256 patches from level 1, (let's assume downscale factors of level 1 relative to level 0 is 4x), then you should specify patch_size to be (1024, 1024) to display the heatmap properly. 

You can use this to display fine attention maps granted you provide the attention scores, and coordinates for overlapping patches. Otherwise, you can also choose to smooth the blocky heatmap using gaussian blur by using blur=True although it won't look as nice.

**Comment by stephandooper** [comment]
Thank you for your answer. I am now able to generate a heatmap that makes a little more sense than before. I found that when using vis_level=-1, the attention map is not scaled and overlayed correctly and you have to set it manually.
I am still struggling with one thing, however. When I try to generate a heatmap, it only seems to output blue colors. 

I trained the models as per the instruction in the README with a single branch for a binary classification problem, and saved the attention scores "A" from logits, Y_prob, Y_hat, A, results_dict = model(data). This gave me a [1, 12745] array which I then passed to visHeatmap. When I only saw outputs with blue colors, I performed a sanity check and went through the model probabilities and chose a slide that contained cancer, which the model also detected with a high confidence (95%). However the output is still blue, and no red to be found. I have attached the output of this particular slide on a blank canvas with its attention values distribution below. 

![download](https://user-images.githubusercontent.com/31624652/111447502-c9820c80-870d-11eb-8edd-699b72343ed6.png)

![attention](https://user-images.githubusercontent.com/31624652/111446360-99863980-870c-11eb-97a5-1534e4c55bea.png)

Is there something I am missing or doing wrong?

P.s. when you say patch coordinates relative to level 0, what representation do you mean? Is it like when level 0 coordinates are (100,100), then the level 1 coordinates should be (25,25) (assuming 4x downsampling)?

**Comment by fedshyvana** [comment]
the coordinates are expected to be always for level 0 (defined by the actual dimensions of the slide) - this is consistent with openslide's read_region convention. So no, if a patch's level 0 coordinates are (100, 100), you would always pass in (100, 100) no matter which level you're trying to access/view that patch. 

As for the blue thingy, did you normalize the attention scores (aka, compress the attention scores to the range [0, 1]) before passing them to visheatmap? If not, you can use convert_to_percentiles=True.

**Comment by fedshyvana** [comment]
btw, thanks for catching vis_level = 1, there is indeed a bug. I just fixed it.

**Comment by stephandooper** [comment]
I think I understand the coordinate system now.

Converting to percentiles did the trick, and it is now displaying attention maps correctly. Thank you again for your help. I will mark this problem as solved and close the issue
